modules/chat/src/routes/routes.ts

```diff
@@ -72,23 +92,84 @@ export class ChatRoutes {
     }
 
     try {
-      await validateUsersInput(this.grpcSdk, users);
+      usersToBeAdded = await validateUsersInput(this.grpcSdk, users);
     } catch (e) {
       throw new GrpcError(status.INTERNAL, e.message);
     }
 
-    room.participants = Array.from(new Set([...room.participants, ...users]));
-    await ChatRoom.getInstance()
-      .findByIdAndUpdate(
-        room._id,
-        room,
-      )
-      .catch((e: Error) => {
-        throw new GrpcError(status.INTERNAL, e.message);
+    const config = await this.grpcSdk.config.get('chat');
+    let retMessage: any [] = [];
+    if (config.sendInvitations.enabled) {
+      const serverConfig = await this.grpcSdk.config.getServerConfig();
+      const url = serverConfig.url;
+      for (const invitedUser of usersToBeAdded) {
+        const invitationsCount = await InvitationToken.getInstance().countDocuments({
+          room: roomId,
+          receiver: invitedUser._id,
+          sender: user._id,
+        });
+        if (room.participants.includes(invitedUser._id)) {
+          retMessage.push({
+            receiver: invitedUser._id,
+            message: 'Already member',
+          });
+          continue;
+        }
+        if (invitationsCount > 0) {
+          retMessage.push({
+            receiver: invitedUser._id,
+            message: 'Already invited',
+          });
+          continue;
+        }
+        let invitationToken: InvitationToken = await InvitationToken.getInstance().create({
+          receiver: invitedUser._id,
+          sender: user._id,
+          token: uuid(),
+          room: roomId,
+        });
+        if (config.sendInvitations.send_email) {
+          let result = { invitationToken, hostUrl: url };
+          const acceptLink = `${result.hostUrl}/hook/chat/accept/${result.invitationToken.token}`;
+          const declineLink = `${result.hostUrl}/hook/chat/decline/${result.invitationToken.token}`;
+          const roomName = room.name;
+          const userName = user.email;
+          await this.emailModule.sendEmail('ChatRoomInvitation', {
+            email: invitedUser.email,
+            sender: 'no-reply',
+            variables: {
+              acceptLink,
+              declineLink,
+              userName,
+              roomName,
+            },
+          });
+        }
+        retMessage.push({
+          receiver: invitedUser._id,
+          message: 'Invitation sent',
+        });
+      }
+      return retMessage;
+    } else {
+      room.participants = Array.from(new Set([...room.participants, ...users]));
```

Comment:
This case could also use a member exists check throwing an error for already included users:
`throw new GrpcError(status.INVALID_ARGUMENT, 'users array contains existing member ids');`

Reply:
^